src/Core/BufferLine.re

```diff
@@ -0,0 +1,152 @@
+/*
+ * Buffer.re
+ *
+ * In-memory text buffer representation
+ */
+module Option = Utility.Option;
+
+exception OutOfBounds;
+
+open CamomileBundled.Camomile;
+let _space = UChar.of_char(' ');
+let tab = UChar.of_char('\t');
+let _cr = UChar.of_char('\r');
+let _lf = UChar.of_char('\n');
+
+type characterCacheInfo = {
+  byteOffset: int,
+  positionOffset: int,
+  uchar: UChar.t,
+  width: int,
+};
+
+let emptyArray = Array.make(0, None);
+
+type t = {
+  indentation: IndentationSettings.t,
+  // [raw] is the raw string (byte array)
+  raw: string,
+  // [characters] is a cache of discovered characters we've found in the string so far
+  mutable characters: array(option(characterCacheInfo)),
+  // nextByte is the nextByte to work from, or -1 if complete
+  mutable nextByte: int,
+  // nextIndex is the nextIndex to work from
+  mutable nextIndex: int,
+  // nextPosition is the graphical position (based on character width)
+  mutable nextPosition: int,
+};
```

Comment:
> This is going to break diffing, which depends on the structural equality of the lines of the new and old documents. 

Ah, good point - we still keep the `raw` string around (and `getLines` actually returns these raw strings - the syntax highlighting still uses it). There still is an extra cost (the `Array.map` to go from `array(BufferLine.t)` -> `array(string)` We actually still need the `array(string)` for SyntaxHighlighting

> Diffing is also currently done every render, although we might eventually want to do it async/time-sliced and only when the document changes.

Definitely! This will be important to maintain perf. We probably only need to recompute the diffs on `BufferUpdate` or when the original resource location changes. And even then - it doesn't  need to be an immediate computation (in other words, it shouldn't block the hot path of key press -> render) - it could be deferred until idle, handed off to be handled asynchronously, or process in time-sliced chunks. That would help keep the core render path as fast as we can.